units/src/amount.rs

```diff
@@ -231,6 +224,50 @@ impl std::error::Error for ParseDenominationError {
     }
 }
 
+/// Parsing error, unknown denomination.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct UnknownDenominationError(InputString);
+
+impl fmt::Display for UnknownDenominationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "unknown denomination {}", self.0.display_cannot_parse("bitcoin denomination"))
```

Comment:
You didn't read the docs, did you? ;)

I think `self.0.unknown_variant("bitcoin denomination")` makes more sense.

Side note: I think we should test `Display` implementations of errors as they are user interface. It'd make it easier to check how they look among preventing accidental changes. (Note that this is not the same as error checking comparing strings!)

Reply:
> You didn't read the docs, did you? ;)

I read them multiple times, I found the whole API hard to understand, as seen by the way I've had two goes at it already, and that was not the first time I looked at that code.

Reply:
OK, sorry, I thought giving a template for the message and an example was completely obvious. What can be improved?

Reply:
Ok, I've attempted to improve the docs, done as a couple of patches at the front of this PR. Can you also please confirm, just to make sure I've got it, that this is the expected usage:

Is this correct expected usage?
```rust
/// An example parsing error.
#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum ParseThingError {
    Invalid(InvalidError),
    Other,
}

impl fmt::Display for ParseThingError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use ParseThingError::*;

        match *self {
            Invalid(ref e) => write_err!(f, "invalid input"; e),
            Other => write!(f, "some other error"),
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for ParseThingError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        use ParseThingError::*;

        match *self {
            Invalid(ref e) => Some(e),
            Other => None,
        }
    }
}

/// Parsing error, unknown denomination.
#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct InvalidError(InputString);

impl fmt::Display for InvalidError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.0.unknown_variant("thing", f)
    }
}

#[cfg(feature = "std")]
impl std::error::Error for InvalidError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
}
```

Reply:
Its the reference to `write_err!` in the docs on `display_cannot_parse` that I don't get.

Reply:
Please see commit `e1c63079 internals: Add docs to InputString` to further determine if I am correctly understanding the API. Thanks man.

Reply:
Yes, that looks correct. The reference to `write_err` was when you have an inner source. e.g.:

```rust
struct FooParseError {
    input: InputString,
    error: core::num::ParseIntError,
}

write_err!(f, "{}", self.input.display_cannot_parse("foo"); self.error)
```

Reply:
Oh sweet, I'll add that to the docs changes. Thanks